engine/flink/kafka-util/src/main/scala/pl/touk/nussknacker/engine/kafka/source/KafkaSourceFactory.scala

```diff
@@ -1,104 +1,76 @@
+import org.apache.flink.types.Nothing
+import org.apache.kafka.clients.consumer.ConsumerRecord
+import pl.touk.nussknacker.engine.api.context.{ProcessCompilationError, ValidationContext}
+import pl.touk.nussknacker.engine.api.context.transformation.{NodeDependencyValue, SingleInputGenericNodeTransformation}
+import pl.touk.nussknacker.engine.api.definition._
+import pl.touk.nussknacker.engine.api.typed.typing.Typed
 
-import javax.validation.constraints.NotBlank
 import scala.reflect.ClassTag
 
-/** <pre>
-  * Wrapper for [[org.apache.flink.streaming.connectors.kafka.FlinkKafkaConsumer]]
+/**
+  * Base factory for Kafka sources with additional metadata variable.
+  * It is based on [[pl.touk.nussknacker.engine.api.context.transformation.SingleInputGenericNodeTransformation]]
+  * that allows custom ValidationContext and Context transformations, which are provided by [[pl.touk.nussknacker.engine.kafka.source.KafkaContextInitializer]]
+ *
+ * Wrapper for [[org.apache.flink.streaming.connectors.kafka.FlinkKafkaConsumer]]
   * Features:
   *   - fetch latest N records which can be later used to test process in UI
   * Fetching data is defined in [[pl.touk.nussknacker.engine.kafka.source.KafkaSource]] which
   * extends [[pl.touk.nussknacker.engine.api.process.TestDataGenerator]]. See [[pl.touk.nussknacker.engine.kafka.KafkaUtils#readLastMessages]]
   *   - reset Kafka's offset to latest value - `forceLatestRead` property, see [[pl.touk.nussknacker.engine.kafka.KafkaUtils#setOffsetToLatest]]
   *
-  * BaseKafkaSourceFactory comes in two variants:
-  *   - KafkaSourceFactory - `topic` parameter has to be passed on frontend
-  *   - SingleTopicKafkaSourceFactory - topic is defined on level of configuration
-  *
-  * </pre>
+  * @param deserializationSchemaFactory - produces KafkaDeserializationSchema for raw [[pl.touk.nussknacker.engine.kafka.source.KafkaSource]]
+  * @param timestampAssigner            - provides timestampAsigner and WatermarkStrategy to KafkaSource
+  * @param formatter                    - support for test data parser and generator
+  * @param processObjectDependencies
+  * @tparam K - type of key of kafka event that is generated by raw source (SourceFunction).
+  * @tparam V - type of value of kafka event that is generated by raw source (SourceFunction).
   * */
-class KafkaSourceFactory[T: ClassTag](deserializationSchemaFactory: KafkaDeserializationSchemaFactory[T],
-                                      timestampAssigner: Option[TimestampWatermarkHandler[T]],
-                                      formatter: RecordFormatter,
-                                      processObjectDependencies: ProcessObjectDependencies)
-  extends BaseKafkaSourceFactory(deserializationSchemaFactory, timestampAssigner, formatter, processObjectDependencies) {
+class KafkaSourceFactory[K: ClassTag, V: ClassTag](deserializationSchemaFactory: KafkaDeserializationSchemaFactory[ConsumerRecord[K, V]],
```

Comment:
I see that after this change we will lost `BaseKafkaSourceFactory`, that had ability to specify more than one topic. Maybe we don't need it anymore but we should think a moment if we really don't need it and if it will be hard to implement it on user side.

Reply:
In our project we already at least have one reference to `BaseKafkaSourceFactory`. Here I assume that multi-topic option can be achieved by implementing proper `override def initialParameters` (and now I see that in `def implementation` I could expose method to extract topics from `params` to allow this multi-topic).

Reply:
More information in `MigrationGuide.md`.